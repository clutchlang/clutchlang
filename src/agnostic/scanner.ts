import { splitLines } from './strings';

/**
 * Represents a section of @member text.
 *
 * Contains enough information in order to represent actionable error or
 * informational messages. Some members in some implementations may be lazy.
 */
export class SourceSpan {
  private static readonly matchNewLine = /\n|\r/;

  constructor(
    public readonly offset: number,
    public readonly column: number,
    public readonly line: number,
    public readonly text: string
  ) {
    if (offset < 0) {
      throw new RangeError(`Invalid offset: ${offset}`);
    }
    if (column < 0) {
      throw new RangeError(`Invalid column: ${column}`);
    }
    if (line < 0) {
      throw new RangeError(`Invalid line: ${line}`);
    }
  }

  /**
   * Length of the span of information.
   */
  public get length(): number {
    return this.text.length;
  }

  /**
   * All lines, separated by line number and text, present in @member text.
   *
   * **NOTE**: `\n\r` is interpreted as a single line break, not two.
   */
  public get lines(): Array<{ line: number; text: string }> {
    return this.isMultiLine
      ? this.computeMultiLines()
      : [
          {
            line: this.line,
            text: this.text,
          },
        ];
  }

  /**
   * Whether the @member text spans over multiple lines.
   *
   * When true, @member lines will represent multiple line numbers.
   */
  public get isMultiLine(): boolean {
    return SourceSpan.matchNewLine.test(this.text);
  }

  private computeMultiLines(): Array<{ line: number; text: string }> {
    return Array.from(
      splitLines(this.text).map((text, index) => {
        return {
          line: this.line + index,
          text,
        };
      })
    );
  }
}

/**
 * A function that handles receiving a @param span and @param error.
 *
 * May optionally return token(s) to recover from the error. Otherwise it is
 * expected the error reporting was fatal and that parsing should be immediately
 * stopped.
 */
export type ErrorReporter = (
  span: SourceSpan,
  error: ScanningError
) => Token | Token[] | void;

/**
 * Error types that may be provided to a @see ErrorReporter.
 */
export type ScanningError = UnexpectedTokenError;

/**
 * A syntax error that occurs as a result of an unexpected @member token.
 *
 * Optionally contains a @member context, or what was being parsed when the
 * error ocurrred, i.e. "argument list" or "function body". May be omitted if
 * the context is not clear or provided.
 */
export class UnexpectedTokenError extends SyntaxError {
  constructor(public readonly token: Token, public readonly context?: string) {
    super(`Unexpected "token"${context ? ` in ${context}` : ''}.`);
  }
}

/**
 * A token that was scanned.
 */
export class Token {}
