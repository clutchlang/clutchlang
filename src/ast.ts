export * from './language/ast/token';
