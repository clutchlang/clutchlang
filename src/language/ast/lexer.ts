export * from './lexer/token';
export * from './lexer/tokenizer';
