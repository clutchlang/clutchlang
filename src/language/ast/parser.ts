export * from './parser/parsers/expression';
export * from './parser/parsers/module';
export * from './parser/parsers/operator';
export * from './parser/parsers/statement';
export * from './parser/factory';
export * from './parser/nodes';
