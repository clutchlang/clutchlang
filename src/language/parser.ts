export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
