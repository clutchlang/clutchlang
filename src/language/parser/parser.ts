import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
