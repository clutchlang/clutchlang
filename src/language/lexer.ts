import { ISourceSpan, StringScanner } from '../agnostic/scanner';
import {
  Characters,
  isDigit,
  isHexadecimal,
  isLetter,
  isWhiteSpace,
} from '../agnostic/strings';

function isIdentifier(character: number): boolean {
  return isIdentifierStart(character) || isDigit(character);
}

function isIdentifierStart(character: number): boolean {
  return isLetter(character) || character === Characters.$_;
}

/**
 * Tokenizes and returns @param program as a series of @see Token.
 */
export function tokenize(program: string, onError?: ErrorReporter): Token[] {
  return new ClutchLexer().scanProgram(new StringScanner(program), onError);
}

/**
 * Reports a @param message at the provided @param span.
 */
export type ErrorReporter = (span: ISourceSpan, error: string) => void;

export enum TokenKind {
  // Literals
  IDENTIFIER = 'Identifier',
  STRING = 'String',
  NUMBER = 'Number',

  // Keywords
  CLASS = 'class',
  ELSE = 'else',
  FALSE = 'false',
  FOR = 'for',
  IF = 'if',
  LET = 'let',
  RETURN = 'return',
  SUPER = 'super',
  THEN = 'then',
  THIS = 'this',
  TRUE = 'true',
  WHILE = 'while',

  // Symbols
  ARROW = '->',
  LEFT_PAREN = '(',
  RIGHT_PAREN = ')',
  LEFT_CURLY = '{',
  RIGHT_CURLY = '}',

  // Operators
  PERIOD = '.',
  PLUS = '+',
  PLUS_EQUALS = '+=',
  MINUS = '-',
  MINUS_EQUALS = '-=',
  STAR = '*',
  STAR_EQUALS = '*=',
  SLASH = '/',
  SLASH_EQUALS = '/=',
  MODULUS = '%',
  MODULUS_EQUALS = '%=',
  EQUALS = '=',
  EQUALS_EQUALS = '==',
  EXCLAIM_EQUALS = '!=',
  EQUALS_EQUALS_EQUALS = '===',
  EXCLAIM_EQUALS_EQUALS = '!==',
  RIGHT_ANGLE = '>',
  RIGHT_ANGLE_EQUALS = '>=',
  LEFT_ANGLE = '<',
  LEFT_ANGLE_EQUALS = '<=',
  PIPE = '|',
  PIPE_PIPE = '||',
  AND = '&',
  AND_AND = '&&',
  CARET = '^',
  EXCLAIM = '!',
  TILDE = '~',
  PLUS_PLUS = '++',
  MINUS_MINUS = '--',
  LEFT_ANGLE_LEFT_ANGLE = '<<',
  RIGHT_ANGLE_RIGHT_ANGLE = '>>',

  // Misc
  EOF = '<EOF>',
}

/**
 * A language-level lexer that yields tokens understood by the grammar.
 */
export class ClutchLexer {
  public static readonly keywords: {
    [index: string]: TokenKind;
  } = {
    class: TokenKind.CLASS,
    else: TokenKind.ELSE,
    false: TokenKind.FALSE,
    for: TokenKind.FOR,
    if: TokenKind.IF,
    let: TokenKind.LET,
    return: TokenKind.RETURN,
    super: TokenKind.SUPER,
    this: TokenKind.THIS,
    true: TokenKind.TRUE,
    while: TokenKind.WHILE,
  };

  private program!: StringScanner;
  private onError!: ErrorReporter;
  private position!: number;
  private lastComments: IComment[] = [];

  public scanProgram(
    program: StringScanner,
    onError: ErrorReporter = (span, message) => {
      throw new SyntaxError(
        `${message} "${span.text}" at ${span.line}:${span.column}`
      );
    }
  ): Token[] {
    this.position = 0;
    this.program = program;
    this.onError = onError;
    const tokens: Token[] = [];
    while (program.hasNext()) {
      const token = this.scanToken();
      if (token) {
        tokens.push(token);
      }
    }
    tokens.push(this.createToken(TokenKind.EOF, ''));
    return tokens;
  }

  private scanToken(): Token | undefined {
    const character = this.program.advance();
    switch (character) {
      case Characters.$LPAREN: // "("
        return this.createToken(TokenKind.LEFT_PAREN);
      case Characters.$RPAREN: // ")"
        return this.createToken(TokenKind.RIGHT_PAREN);
      case Characters.$LCURLY: // "{"
        return this.createToken(TokenKind.LEFT_CURLY);
      case Characters.$RCURLY: // "}"
        return this.createToken(TokenKind.RIGHT_CURLY);
      case Characters.$PERIOD: // "."
        return this.createToken(TokenKind.PERIOD);
      case Characters.$PLUS: // "+" or "+=" or "++"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.PLUS_EQUALS
            : this.program.match(Characters.$PLUS)
              ? TokenKind.PLUS_PLUS
              : TokenKind.PLUS
        );
      case Characters.$MINUS: // "-" or "--" or "-=" or "->"
        return this.createToken(
          this.program.match(Characters.$RANGLE)
            ? TokenKind.ARROW
            : this.program.match(Characters.$EQUALS)
              ? TokenKind.MINUS_EQUALS
              : this.program.match(Characters.$MINUS)
                ? TokenKind.MINUS_MINUS
                : TokenKind.MINUS
          // Code coverage considers the `);` to be a missed line :(
          /* istanbul ignore next */
        );
      case Characters.$STAR: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.STAR_EQUALS
            : TokenKind.STAR
        );
      case Characters.$PERCENT: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.MODULUS_EQUALS
            : TokenKind.MODULUS
        );
      case Characters.$EQUALS: // "=" or "==" or "==="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EQUALS_EQUALS_EQUALS
              : TokenKind.EQUALS_EQUALS
            : TokenKind.EQUALS
        );
      case Characters.$LANGLE: // "<" or "<=" or "<<"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.LEFT_ANGLE_EQUALS
            : this.program.match(Characters.$LANGLE)
              ? TokenKind.LEFT_ANGLE_LEFT_ANGLE
              : TokenKind.LEFT_ANGLE
        );
      case Characters.$RANGLE: // ">" or ">=" or ">>"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.RIGHT_ANGLE_EQUALS
            : this.program.match(Characters.$RANGLE)
              ? TokenKind.RIGHT_ANGLE_RIGHT_ANGLE
              : TokenKind.RIGHT_ANGLE
        );
      case Characters.$SLASH: // "/" or "//" or "/="
        return this.scanSlash();
      case Characters.$SQUOTE: // "'"
        return this.scanString();
      case Characters.$PIPE: // "|" or "||"
        return this.createToken(
          this.program.match(Characters.$PIPE)
            ? TokenKind.PIPE_PIPE
            : TokenKind.PIPE
        );
      case Characters.$AND: // "&" or "&&"
        return this.createToken(
          this.program.match(Characters.$AND)
            ? TokenKind.AND_AND
            : TokenKind.AND
        );
      case Characters.$TILDE: // "~"
        return this.createToken(TokenKind.TILDE);
      case Characters.$CARET: // "^"
        return this.createToken(TokenKind.CARET);
      case Characters.$EXCLAIM: // "!" or "!=" or "!==""
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EXCLAIM_EQUALS_EQUALS
              : TokenKind.EXCLAIM_EQUALS
            : TokenKind.EXCLAIM
        );
      default:
        if (isWhiteSpace(character)) {
          this.position = this.program.position;
          break;
        }
        if (isDigit(character)) {
          return this.scanNumber(character);
        }
        if (isIdentifierStart(character)) {
          return this.scanIdentifierOrKeyword();
        }
        this.reportError(
          `Unexpected character "${String.fromCharCode(character)}"`
        );
    }
    return;
  }

  /**
   * Scans a string, returning `undefined` if there was no string.
   */
  private scanString(): Token {
    while (this.program.hasNext()) {
      if (this.program.match(Characters.$SQUOTE)) {
        this.position++;
        let lexeme = this.substring();
        lexeme = lexeme.substring(0, lexeme.length - 1);
        return this.createToken(TokenKind.STRING, lexeme);
      }
      this.program.advance();
    }
    this.reportError('Unterminated string');
    this.position++;
    return this.createToken(TokenKind.STRING);
  }

  /**
   * Scans any tokens that are valid numeric literals
   */
  private scanNumber(starting: number): Token {
    if (starting === Characters.$0) {
      if (
        this.program.match(Characters.$X) ||
        this.program.match(Characters.$x)
      ) {
        return this.scanHexNumber();
      }
    }
    if (
      this.program.match(Characters.$E) ||
      this.program.match(Characters.$e)
    ) {
      return this.scanDigits();
    }
    while (this.program.hasNext()) {
      const next = this.program.peek();
      if (this.program.match(Characters.$PERIOD)) {
        break;
      }
      if (!isDigit(next)) {
        break;
      }
      this.program.advance();
    }
    return this.scanDigits();
  }

  /**
   * Scans any tokens that are valid hexadecimel digits, returning a number.
   */
  private scanHexNumber(): Token {
    this.scanPredicate(isHexadecimal);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid digits, returning a number.
   */
  private scanDigits(): Token {
    this.scanPredicate(isDigit);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid identifiers.
   */
  private scanIdentifierOrKeyword(): Token {
    this.scanPredicate(isIdentifier);
    const identifierOrKeyword = this.substring();
    return new Token(
      ClutchLexer.keywords[identifierOrKeyword] || TokenKind.IDENTIFIER,
      identifierOrKeyword,
      this.clearComments(),
      this.position - identifierOrKeyword.length
    );
  }

  /**
   * Scans tokens preceding with "/".
   */
  private scanSlash(): Token | undefined {
    // "//"
    if (this.program.match(Characters.$SLASH)) {
      this.advanceToEOL();
      this.lastComments.push({
        lexeme: this.substring().trim(),
        offset: this.position,
      });
      return;
    }
    // "/" or "/="
    return this.createToken(
      this.program.match(Characters.$EQUALS)
        ? TokenKind.SLASH_EQUALS
        : TokenKind.SLASH
    );
  }

  /**
   * Scans until the EOF or until @param predicate returns false.
   */
  private scanPredicate(predicate: (character: number) => boolean): void {
    while (this.program.hasNext()) {
      const peek = this.program.peek();
      if (!predicate(peek)) {
        break;
      }
      this.program.advance();
    }
  }

  /**
   * Advances the position of the scanner until either EOF or EOL.
   *
   * **NOTE**: Treats `CR+LF` as a DOS-stlye line ending, not two lines.
   */
  private advanceToEOL(): void {
    while (this.program.hasNext()) {
      // "\n" (LF)
      if (this.program.match(Characters.$LF)) {
        return;
      }
      // "\r" (CR)
      if (this.program.match(Characters.$CR)) {
        // "\r\n" (CR+LF)
        this.program.match(Characters.$LF);
        return;
      }
      this.program.advance();
    }
  }

  /**
   * Returns a new token with a @param kind and the current substring.
   */
  private createToken(kind: TokenKind, content = this.substring()): Token {
    return new Token(
      kind,
      content,
      this.clearComments(),
      this.position - content.length
    );
  }

  /**
   * Clear and retrieve comments lexed.
   */
  private clearComments(): IComment[] {
    const comments = this.lastComments;
    this.lastComments = [];
    return comments;
  }

  /**
   * Returns the current substring, advancing the position counter.
   */
  private substring(): string {
    const start = this.position;
    const end = this.program.position;
    this.position = end;
    return this.program.substring(start, end);
  }

  /**
   * Reports a lexical @param error occurring at @param offset.
   */
  private reportError(error: string, offset = this.position): void {
    this.onError(this.program.source.span(offset, offset + 1), error);
  }
}

/**
 * A scanned token from @function tokenize.
 */
export interface IToken {
  readonly kind: TokenKind;
  readonly lexeme: string;
  readonly comments: IComment[];
  readonly offset: number;
}

/**
 * Implementation of @interface IToken private to @function tokenize.
 */
class Token implements IToken {
  constructor(
    public readonly kind: TokenKind,
    public readonly lexeme: string,
    public readonly comments: IComment[],
    public readonly offset: number
  ) {}
}

export interface IComment {
  readonly lexeme: string;
  readonly offset: number;
}
