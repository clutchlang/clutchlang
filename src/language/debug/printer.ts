import { StringBuffer } from "../../agnostic/strings";
import * as ast from "../ast/parser";

export class PrintTreeVisitor extends ast.AstVisitor<StringBuffer, StringBuffer> {
  constructor(private readonly indent = 2) {
    super();
  }

  public visitBinaryExpression(
    node: ast.BinaryExpression, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      'BinaryExpression',
      buffer, [
        ['Left', node.left],
        ['Operator', node.operator],
        ['Right', node.right],
      ],
    );
  }

  public visitCallExpression(
    node: ast.CallExpression<ast.Expression>, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      'CallExpression',
      buffer, [
        ['Target', node.target],
        ['Arguments', node.args],
      ],
    );
  }
  
  public visitConditionalExpression(
    node: ast.ConditionalExpression, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      'ConditionalExpression',
      buffer, [
        ['If', node.condition],
        ['Then', node.body],
        ['Else', node.elseBody],
      ],
    );
  }

  public visitFunctionDeclaration(
    node: ast.FunctionDeclaration, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      'FunctionDeclaration',
      buffer, [
        ['Name', node.name],
        ['External', node.isExternal],
        ['Parameters', node.params],
        ['Returns', node.returnType],
        ['Body', node.body],
      ],
    );
  }

  public visitGroupExpression(
    node: ast.GroupExpression<ast.Expression>, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      'GroupExpression',
      buffer, [
        ['Expression', node.expression],
      ],
    );
  } 

  public visitIdentifier(
    node: ast.Identifier, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `Identifier {${node.name}}`,
      buffer,
    );
  }

  public visitLiteralBoolean(
    node: ast.LiteralBoolean, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `LiteralBoolean {${node.value}}`,
      buffer,
    );
  }

  public visitLiteralNumber(
    node: ast.LiteralNumber, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `LiteralNumber {${node.value}}`,
      buffer,
    );
  }
  
  public visitLiteralString(
    node: ast.LiteralString, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `LiteralString {${node.value}}`,
      buffer,
    );
  }

  public visitModuleDeclaration(
    node: ast.ModuleDeclaration, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `ModuleDeclaration`,
      buffer,
      [
        ['Declarations', node.declarations]
      ],
    );
  }

  public visitModuleRoot(
    node: ast.ModuleRoot, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `ModuleRoot`,
      buffer,
      [
        ['Modules', node.modules]
      ],
    );
  }

  public visitOperator(
    node: ast.Operator<ast.OperatorType>, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `Operator {${node.type}: "${node.firstToken.lexeme}"}`,
      buffer,
    );
  }

  public visitParameterList(
    node: ast.ParameterList, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `ParameterList`,
      buffer,
      [
        ['Parameters', node.params]
      ],
    );
  }

  public visitPrefixExpression(
    node: ast.PrefixExpression, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `PrefixExpression`,
      buffer,
      [
        ['Operator', node.operator],
        ['Target', node.target]
      ],
    );
  }

  public visitPostfixExpression(
    node: ast.PostfixExpression, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `PostfixExpression`,
      buffer,
      [
        ['Target', node.target],
        ['Operator', node.operator],
      ],
    );
  }

  public visitPropertyExpression(
    node: ast.PropertyExpression<ast.Expression>, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `PropertyExpression`,
      buffer,
      [
        ['Target', node.target],
        ['Property', node.property],
      ],
    );
  }

  public visitReturnStatement(
    node: ast.ReturnStatement, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `ReturnStatement`,
      buffer,
      [
        ['Expression', node.expression],
      ],
    );
  }

  public visitStatementBlock(
    node: ast.StatementBlock, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `StatementBlock`,
      buffer,
      [
        ['Statements', node.statements],
      ],
    );
  }

  public visitTypeDeclaration(
    node: ast.TypeDeclaration, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `TypeDeclaration`,
      buffer,
      [
        ['Name', node.name],
        ['External', node.isExternal],
        ['Members', node.members]
      ],
    );
  }

  public visitVariableDeclaration(
    node: ast.VariableDeclaration, 
    buffer = new StringBuffer(),
  ): StringBuffer {
    return this.visitAll(
      `VariableDeclaration`,
      buffer,
      [
        ['Name', node.name],
        ['Type', node.type],
        ['Value', node.value]
      ],
    );
  }

  private visitAll(
    name: string, 
    buffer: StringBuffer, 
    children: ReadonlyArray<[string, ast.AstNode | ast.AstNode[] | undefined | boolean | string]> = [],
  ): StringBuffer {
    buffer.writeLine(`${name}:`);
    for (const child of children) {
      const title = child[0];
      const value = child[1];
      if (value === undefined) {
        continue;
      }
      if (typeof value === 'boolean' || typeof value === 'string') {
        buffer.writeLine(`${title}: ${value}`);
        continue;
      }
      buffer.writeLine(`${title}:`);
      if (value instanceof Array) {
        value.forEach(n => this.visitWithIndent(n, buffer));
      } else {
        this.visitWithIndent(value, buffer);
      }
    }
    return buffer;
  }

  private visitWithIndent(node: ast.AstNode, buffer: StringBuffer): void {
    buffer.indent(this.indent);
    node.accept(this, buffer);
    buffer.indent(-this.indent);
  }
}
