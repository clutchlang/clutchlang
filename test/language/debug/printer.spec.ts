// tslint:disable:no-magic-numbers

import { SourceFile } from '../../../src/agnostic/scanner';
import * as lexer from '../../../src/language/ast/lexer';
import {
  StaticMessageCode,
  StaticMessageReporter,
} from '../../../src/language/ast/message';
import * as ast from '../../../src/language/ast/parser';
import { PrintTreeVisitor } from '../../../src/language/debug/printer';

function parseModuleRoot(text: string): ast.ModuleRoot {
  // TODO: Move the following block into a common test-infra area.
  const source = new SourceFile(text, 'printer.spec.ts');
  const reporter = new StaticMessageReporter(source);
  const tokens = lexer.tokenize(text, (offset, length) => {
    reporter.reportOffset(
      offset,
      length,
      StaticMessageCode.SYNTAX_UNEXPECTED_TOKEN
    );
  });
  const parser = new ast.ModuleParser(tokens, reporter);
  return parser.parseModuleRoot();
}

test('should print expressions', () => {
  expect(parseModuleRoot(`
    main -> {
      // Binary
      1 + 1
      // Call
      call()
      // Conditional
      if a then b else c
      // Group
      (1 + 1)
      // Literals
      fooBar
      true
      false
      42
      'Hello World'
      // Prefix
      -1
      // Postfix
      n++
      // Property
      a.b

    }
  `).accept(new PrintTreeVisitor())).toMatchSnapshot();
});

test('should print other nodes', () => {
  expect(parseModuleRoot(`
    type Foo {
      bar -> {
        return true
      }
      baz -> {
        return
      }
    }

    let x: String = 'Hello'
  `).accept(new PrintTreeVisitor())).toMatchSnapshot();
});
