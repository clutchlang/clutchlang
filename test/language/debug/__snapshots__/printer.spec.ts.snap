// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print expressions 1`] = `
StringBuffer {
  "buffer": "ModuleRoot:
Modules:
  ModuleDeclaration:
  Declarations:
    FunctionDeclaration:
    Name:
      Identifier {main}:
    External: false
    Body:
      StatementBlock:
      Statements:
        BinaryExpression:
        Left:
          LiteralNumber {1}:
        Operator:
          Operator {13: \\"+\\"}:
        Right:
          LiteralNumber {1}:
        CallExpression:
        Target:
          Identifier {call}:
        Arguments:
        ConditionalExpression:
        If:
          Identifier {a}:
        Then:
          Identifier {b}:
        Else:
          CallExpression:
          Target:
            Identifier {c}:
          Arguments:
            BinaryExpression:
            Left:
              LiteralNumber {1}:
            Operator:
              Operator {13: \\"+\\"}:
            Right:
              LiteralNumber {1}:
        Identifier {fooBar}:
        LiteralBoolean {true}:
        LiteralBoolean {false}:
        LiteralNumber {42}:
        BinaryExpression:
        Left:
          LiteralString {Hello World}:
        Operator:
          Operator {14: \\"-\\"}:
        Right:
          LiteralNumber {1}:
        PostfixExpression:
        Target:
          Identifier {n}:
        Operator:
          Operator {6: \\"++\\"}:
        PropertyExpression:
        Target:
          Identifier {a}:
        Property:
          Identifier {b}:
",
  "indents": "",
}
`;
